8352587: C2 SuperWord: we must avoid Multiversioning for PeelMainPost loops

eme64 opened the pull request:
This was a fuzzer failure, which hit an assert in SuperWord:

`# assert(_cl->is_multiversion_fast_loop() == (_multiversioning_fast_proj != nullptr)) failed: must find the multiversion selector IFF loop is a multiversion fast loop`

We had a fast main loop, but it could not find the `multiversion_if`. The reason was that the loop was a `PeelMainPost` loop, i.e. there is no pre-loop but only a single peeled iteration. This makes the pattern matching from main-loop via pre-loop to `multiversion_if` impossible.

I'm proposing two changes in this PR:
- We must check `peel_only`, to see if we are in a `PeelMainPost` or `PreMainPost` case, and only do multiversioning if we know that there will be a pre-loop.
- In `eliminate_useless_multiversion_if` we should already detect that a main-loop that is marked as multiversioned should be able to find its `multiversion_if`. I'm removing its multiversioning marking if we cannot find the `multiversion_if`.

I added 2 tests:
- The fuzzer generated test that hits the assert before this patch.
- An IR test that checks that we do not multiversion in a  `PeelMainPost` loop case.

---------------

**FYI**: I tried to add an assert in `eliminate_useless_multiversion_if` that we must always find the `multiversion_if` from a multiversioned main loop. But there are cases where this can fail. Here an example:

`test/hotspot/jtreg/compiler/locks/TestSynchronizeWithEmptyBlock.java`

With flags: `-ea -esa -XX:CompileThreshold=100 -XX:+UnlockExperimentalVMOptions -server -XX:-TieredCompilation`

```
Counted            Loop: N537/N176  counted [int,100),+1 (-1 iters) 
Loop: N0/N0  has_sfpt
  Loop: N307/N361  limit_check profile_predicated predicated sfpts={ 182 495 }
    Loop: N536/N535 
      Loop: N537/N176  counted [int,100),+1 (-1 iters)  has_sfpt strip_mined
    Loop: N379/N383  limit_check profile_predicated predicated counted [int,int),+1 (4 iters)  pre rc  has_sfpt
    Loop: N353/N357  counted [int,1000),+1 (4 iters)  post rc  has_sfpt
Multiversion       Loop: N537/N176  counted [int,100),+1 (100 iters)  has_sfpt strip_mined
PreMainPost        Loop: N537/N176  counted [int,100),+1 (100 iters)  multiversion_fast has_sfpt strip_mined
Unroll 2           Loop: N537/N176  counted [int,100),+1 (100 iters)  main multiversion_fast has_sfpt strip_mined
Poor node estimate: 306 >> 92
Loop: N0/N0  has_sfpt
  Loop: N307/N361  limit_check profile_predicated predicated sfpts={ 182 }
    Loop: N556/N557  sfpts={ 559 }
      Loop: N552/N554  counted [int,100),+1 (100 iters)  multiversion_delayed_slow has_sfpt strip_mined
    Loop: N599/N601  counted [int,int),+1 (4 iters)  pre multiversion_fast has_sfpt
    Loop: N536/N535  sfpts={ 538 }
      Loop: N629/N176  counted [int,99),+2 (100 iters)  main multiversion_fast has_sfpt strip_mined
    Loop: N575/N577  counted [int,100),+1 (4 iters)  post multiversion_fast has_sfpt
    Loop: N379/N383  limit_check profile_predicated predicated counted [int,int),+1 (4 iters)  pre rc  has_sfpt
    Loop: N353/N357  counted [int,1000),+1 (4 iters)  post rc  has_sfpt
Parallel IV: 643       Loop: N552/N554  counted [int,100),+1 (100 iters)  multiversion_delayed_slow has_sfpt strip_mined
Parallel IV: 646     Loop: N599/N601  counted [int,int),+1 (4 iters)  pre multiversion_fast has_sfpt
Parallel IV: 652       Loop: N629/N176  counted [int,99),+2 (100 iters)  main multiversion_fast has_sfpt strip_mined
Parallel IV: 649     Loop: N575/N577  counted [int,100),+1 (4 iters)  post multiversion_fast has_sfpt
Loop: N0/N0  has_sfpt
  Loop: N307/N361  limit_check profile_predicated predicated sfpts={ 182 }
    Loop: N556/N557  sfpts={ 559 }
      Loop: N552/N554  counted [int,100),+1 (100 iters)  multiversion_delayed_slow has_sfpt strip_mined
    Loop: N599/N601  counted [int,int),+1 (4 iters)  pre multiversion_fast has_sfpt
    Loop: N536/N535  sfpts={ 538 }
      Loop: N629/N176  counted [int,99),+2 (100 iters)  main multiversion_fast has_sfpt strip_mined
    Loop: N575/N577  counted [int,100),+1 (4 iters)  post multiversion_fast has_sfpt
    Loop: N379/N383  limit_check profile_predicated predicated counted [int,int),+1 (4 iters)  pre rc  has_sfpt
    Loop: N353/N357  counted [int,1000),+1 (4 iters)  post rc  has_sfpt
Empty without zero trip guard         Loop: N552/N554  counted [int,100),+1 (100 iters)  multiversion_delayed_slow has_sfpt strip_mined
Peel               Loop: N552/N554  counted [int,100),+1 (100 iters)  multiversion_delayed_slow has_sfpt strip_mined
Empty without zero trip guard       Loop: N599/N601  counted [int,int),+1 (4 iters)  pre multiversion_fast has_sfpt
Peel             Loop: N599/N601  counted [int,int),+1 (4 iters)  pre multiversion_fast has_sfpt
Unroll 4           Loop: N629/N176  counted [int,99),+2 (100 iters)  main multiversion_fast has_sfpt strip_mined
```

It seems that we are able to detect some loops as empty loops, including the pre-loop. But somhow the main-loop is not removed by "empty loop", and now this main-loop cannot traverse through the pre-loop to the `multiversion_if`.

If reviewers thing this really should be investigated, I could file a follow-up RFE.

Update: I filed:
[JDK-8352819](https://bugs.openjdk.org/browse/JDK-8352819) C2 SuperWord: add assert to eliminate_useless_multiversion_if

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8352587](https://bugs.openjdk.org/browse/JDK-8352587): C2 SuperWord: we must avoid Multiversioning for PeelMainPost loops (**Bug** - P3)


### Reviewers
 * [Vladimir Kozlov](https://openjdk.org/census#kvn) (@vnkozlov - **Reviewer**) Review applies to [1ab4f731](https://git.openjdk.org/jdk/pull/24183/files/1ab4f731c7e7322949c9236b2a338f441c3d4d14)
 * [Christian Hagedorn](https://openjdk.org/census#chagedorn) (@chhagedorn - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/24183/head:pull/24183` \
`$ git checkout pull/24183`

Update a local copy of the PR: \
`$ git checkout pull/24183` \
`$ git pull https://git.openjdk.org/jdk.git pull/24183/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 24183`

View PR using the GUI difftool: \
`$ git pr show -t 24183`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/24183.diff">https://git.openjdk.org/jdk/pull/24183.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/24183#issuecomment-2748272450)
</details>

vnkozlov commented:
> If reviewers thing this really should be investigated, I could file a follow-up RFE.

Yes, please.  Can you check without "strip minning" if we can eliminate this loop?

vnkozlov reviewed (approved):
Looks reasonable.

chhagedorn commented on test/hotspot/jtreg/compiler/loopopts/superword/TestMultiversionWithPeelMainPost.java:
```diff
@@ -0,0 +1,58 @@
+ * @bug 8352587
+ * @summary Test case where we used to Multiversion a PeelMainPost loop,
+ *          which is useless and triggered an assert later on.
+ * @run driver compiler.loopopts.superword.TestMultiversionWithPeelMainPost
```
Should be `main` to be able to add flags in the CI:
```suggestion
 * @run main compiler.loopopts.superword.TestMultiversionWithPeelMainPost
```

chhagedorn commented on test/hotspot/jtreg/compiler/loopopts/superword/TestPeelMainPostNoMultiversioning.java:
```diff
@@ -0,0 +1,67 @@
+        long x = value;
+        long y = multiplicator;
+        for (int i = 0; i < 10_000; i++) {
+            x *= y; // No memory load/stroe -> PeelMainPost
```
```suggestion
            x *= y; // No memory load/store -> PeelMainPost
```

chhagedorn commented on src/hotspot/share/opto/loopTransform.cpp:
```diff
@@ -3430,10 +3430,15 @@ bool IdealLoopTree::iteration_split_impl(PhaseIdealLoop *phase, Node_List &old_n
-      phase->maybe_multiversion_for_auto_vectorization_runtime_checks(this, old_new);
+      if (!peel_only) {
+        // We are going to add pre-loop and post-loop (PreMainPost).
+        // But should we also multi-version for auto-vectorization speculative
```
Just a side note: It seems that we sometimes say "multiversion" and sometimes "multi-version". I find the latter more readable. Maybe we can make it consistent at some point. This could also mean that we have `*_multi_version_*` in method names instead of `*_multiversion*_`.

eme64 replied:
Hmm... good catch. Personally, I prefer "multiversion" as a single word. But this is very subjective :sweat_smile:

eme64 replied:
I now renamed the 2 occurances of `multi-version` to `multiversion`, since that was a minimal diff. We could still file an RFE to rename everything later.

chhagedorn replied:
Your call on what you like better since you introduced the concept :-) It should just be consistent.

eme64 commented:
@vnkozlov Thanks for your review!

@chhagedorn I addressed all your comments :)

I filed:
[JDK-8352819](https://bugs.openjdk.org/browse/JDK-8352819) C2 SuperWord: add assert to eliminate_useless_multiversion_if

chhagedorn reviewed (approved):
Update looks good!

eme64 commented:
@vnkozlov @chhagedorn Thanks for the reviews!

/integrate
